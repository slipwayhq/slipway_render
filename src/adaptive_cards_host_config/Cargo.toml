[package]
name = "adaptive_cards_host_config"
version.workspace = true
edition.workspace = true
authors.workspace = true

[dependencies]
serde = { workspace = true, features = ["derive"] }

[build-dependencies]
populate_schema_defaults = { path = "../populate_schema_defaults" }
