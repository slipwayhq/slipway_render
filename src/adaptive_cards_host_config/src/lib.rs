#[allow(
    clippy::to_string_trait_impl,
    clippy::derivable_impls,
    clippy::wrong_self_convention
)]
mod generated;

pub use generated::*;

pub fn default() -> HostConfig {
    crate::builder::HostConfig::default()
        .try_into()
        .expect("default host config should be valid")
}

impl HostConfig {
    pub fn get_all_font_family_stacks(&self) -> Vec<&str> {
        vec![
            &self.font_types.default.font_family,
            &self.font_types.monospace.font_family,
        ]
    }
}
