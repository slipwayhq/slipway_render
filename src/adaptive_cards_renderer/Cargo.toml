[package]
name = "adaptive_cards_renderer"
version.workspace = true
edition.workspace = true
authors.workspace = true

[dependencies]
adaptive_cards = { path = "../adaptive_cards" }
adaptive_cards_host_config = { path = "../adaptive_cards_host_config" }
csscolorparser = "0.7.0"
image.workspace = true
imageproc = "0.25.0"
serde = { workspace = true, features = ["derive"] }
serde_json = { workspace = true }
taffy = "0.5.2"
thiserror = "1.0.63"
parley = { version = "0.2.0", default-features = false, features = [
    "libm",
    "std",
] }
skrifa = { version = "0.22.3" }
peniko = { version = "0.2.0" }
swash = { version = "0.1.19", default-features = true }

[build-dependencies]
adaptive_cards_types_generator = { path = "../adaptive_cards_types_generator" }
populate_schema_defaults = { path = "../populate_schema_defaults" }

[dev-dependencies]
walkdir = "2.5.0"
jsonschema = { workspace = true }
insta = { workspace = true, features = ["json", "glob"] }
