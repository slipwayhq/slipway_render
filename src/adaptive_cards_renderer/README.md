# Overview

## Type Generation

We generate custom types from the Adaptive Cards typed schema files located in
`adaptive_cards_data/schema/typed-schema-1.6`.

Every Adaptive Cards type must implement the `Layoutable` trait. The code generator adds the
default trait implementation to any types marked as not explicitly implemented (see the
`UNIMPLEMENTED_LAYOUTABLE_TYPES` array).

As we implement support rendering more Adaptive Cards types we will remove the types from
the unimplemented list and implement the trait explicitly.

The code generator also implements the `HasLayoutData` trait for each type and adds the required 
`RefCell<ElementLayoutData>` data to the struct.

The `ElementLayoutData` contains the transient data used during the layout and draw passes.

The default `Layoutable` trait implementations of `layout` and `draw` are designed to be sufficient
for all types, and instead each type should override the implementation of `layout_override` and
`draw_override` which are called by `layout` and `draw` respectively.

Rendering starts with the `crate::render::render` method. This performs the card's `layout` pass which
populates all the data required by Taffy, then asks Taffy to compute the layout, and finally creates
the output image and asks the card to draw itself to the image.