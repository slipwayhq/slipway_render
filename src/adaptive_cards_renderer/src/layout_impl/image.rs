use std::{cell::RefCell, rc::Rc};

use adaptive_cards::{BlockElementHeight, ImageSize, ImageStyle, StringOrBlockElementHeight};
use image::{RgbaImage, imageops::FilterType};
use imageproc::{
    drawing::{Canvas, draw_filled_rect_mut},
    point::Point,
    rect::Rect,
};
use taffy::{Dimension, Style, TaffyTree};

use crate::{
    ElementLayoutData,
    element_layout_data::ElementTaffyData,
    errors::{RenderError, TaffyErrorToRenderError},
    host_config_utils::StringToColor,
    layout_context::LayoutContext,
    layout_scratch::LayoutScratch,
    masked_image::MaskedImage,
    utils::{ClampToI32, ClampToU32, TaffyLayoutUtils},
};

use super::{
    image_node_context::ImageNodeContext,
    measure::NodeContext,
    utils::{apply_horizontal_alignment, parse_dimension},
};

enum SourceImageSize {
    Fixed,
    Dynamic,
}

impl SourceImageSize {
    fn from_image(image: &adaptive_cards::Image<ElementLayoutData>) -> Self {
        if image.url.contains("$width") || image.url.contains("$height") {
            Self::Dynamic
        } else {
            Self::Fixed
        }
    }
}

impl crate::layoutable::Layoutable for adaptive_cards::Image<ElementLayoutData> {
    fn layout_override(
        &self,
        context: &LayoutContext,
        baseline_style: Style,
        tree: &mut TaffyTree<NodeContext>,
        _scratch: &mut LayoutScratch,
    ) -> Result<ElementTaffyData, RenderError> {
        // Handled by parent
        // self.is_visible
        // self.separator
        // self.spacing
        // self.height

        let mut style = Style { ..baseline_style };
        apply_horizontal_alignment(self.horizontal_alignment, &mut style, context);

        let size = self.size.unwrap_or(ImageSize::Auto);

        if let Some(width) = self.width.as_ref() {
            style.size.width = parse_dimension(width, context)?;
        } else {
            match size {
                adaptive_cards::ImageSize::Auto => {
                    // Image will scale down to fit if needed, but will not scale up to fill the area.
                    style.max_size.width = Dimension::Percent(100.);
                    style.max_size.height = Dimension::Percent(100.);
                }
                adaptive_cards::ImageSize::Stretch => {
                    // Image with both scale down and up to fit as needed.
                    style.max_size.width = Dimension::Percent(100.);
                    style.max_size.height = Dimension::Percent(100.);
                }
                adaptive_cards::ImageSize::Small => {
                    style.size.width =
                        Dimension::Length(context.host_config.image_sizes.small as f32);
                }
                adaptive_cards::ImageSize::Medium => {
                    style.size.width =
                        Dimension::Length(context.host_config.image_sizes.medium as f32);
                }
                adaptive_cards::ImageSize::Large => {
                    style.size.width =
                        Dimension::Length(context.host_config.image_sizes.large as f32);
                }
            }
        }

        let source_image_size = SourceImageSize::from_image(self);
        match source_image_size {
            SourceImageSize::Dynamic => {
                if let StringOrBlockElementHeight::BlockElementHeight(BlockElementHeight::Auto) =
                    self.height
                {
                    context.host_context.log_warn(&format!(
                        "Image height is set to Auto, but a dynamically sized image is present. This will likely result in a zero height image. Path: {}",
                        context.path
                    ));
                }

                tree.new_leaf(style)
                    .err_context(context)
                    .map(ElementTaffyData::from)
            }
            SourceImageSize::Fixed => {
                let source_image = load_source_image(self, context, None);

                let source_width = source_image.width() as f32;
                let source_height = source_image.height() as f32;

                let max_width = if self.width.as_ref().is_some() {
                    None
                } else {
                    match size {
                        adaptive_cards::ImageSize::Auto => Some(source_width),
                        adaptive_cards::ImageSize::Stretch => None,
                        adaptive_cards::ImageSize::Small => {
                            Some(context.host_config.image_sizes.small as f32)
                        }
                        adaptive_cards::ImageSize::Medium => {
                            Some(context.host_config.image_sizes.medium as f32)
                        }
                        adaptive_cards::ImageSize::Large => {
                            Some(context.host_config.image_sizes.large as f32)
                        }
                    }
                };

                let image_context = ImageNodeContext {
                    source_width,
                    source_height,
                    max_width,
                };

                let node_context = NodeContext::Image(image_context);
                tree.new_leaf_with_context(style, node_context)
                    .err_context(context)
                    .map(ElementTaffyData::from)
            }
        }
    }

    fn draw_override(
        &self,
        context: &LayoutContext,
        tree: &TaffyTree<NodeContext>,
        taffy_data: &ElementTaffyData,
        image: Rc<RefCell<MaskedImage>>,
        _scratch: &mut LayoutScratch,
    ) -> Result<(), RenderError> {
        let node_layout = tree.layout(taffy_data.node_id).err_context(context)?;
        let absolute_rect = node_layout.absolute_rect(context);

        let size = self.size.unwrap_or(ImageSize::Auto);

        let mut source_image = load_source_image(self, context, Some(absolute_rect));

        let (source_image_width, source_image_height) = source_image.dimensions();
        match size {
            ImageSize::Auto => {
                // Image will scale down to fit if needed, but will not scale up to fill the area.
                if source_image_width > absolute_rect.width()
                    || source_image_height > absolute_rect.height()
                {
                    source_image = scale_to_fit_rect(
                        &source_image,
                        absolute_rect.width(),
                        absolute_rect.height(),
                    );
                }
            }
            ImageSize::Stretch => {
                // Image with both scale down and up to fit as needed.
                if source_image_width != absolute_rect.width()
                    && source_image_height != absolute_rect.height()
                {
                    source_image = scale_to_fit_rect(
                        &source_image,
                        absolute_rect.width(),
                        absolute_rect.height(),
                    );
                }
            }
            ImageSize::Small => {
                let target_width = context.host_config.image_sizes.small.clamp_to_u32();
                scale_to_fit_width(
                    source_image_width,
                    target_width,
                    source_image_height,
                    &mut source_image,
                );
            }
            ImageSize::Medium => {
                let target_width = context.host_config.image_sizes.medium.clamp_to_u32();
                scale_to_fit_width(
                    source_image_width,
                    target_width,
                    source_image_height,
                    &mut source_image,
                );
            }
            ImageSize::Large => {
                let target_width = context.host_config.image_sizes.large.clamp_to_u32();
                scale_to_fit_width(
                    source_image_width,
                    target_width,
                    source_image_height,
                    &mut source_image,
                );
            }
        }

        let horizontal_alignment = self
            .horizontal_alignment
            .unwrap_or(context.inherited.horizontal_alignment);

        let final_position =
            get_new_absolute_position(absolute_rect, &source_image, horizontal_alignment);
        let final_rect = Rect::at(final_position.x, final_position.y)
            .of_size(source_image.width(), source_image.height());

        let style = self.style.unwrap_or(ImageStyle::Default);
        match style {
            ImageStyle::Default => {
                let mut image_mut = image.borrow_mut();
                if let Some(background_color) = self.background_color.as_ref() {
                    let background_color = background_color.to_color()?;
                    draw_filled_rect_mut(&mut *image_mut, final_rect, background_color);
                }
                copy_image(&source_image, &mut image_mut, final_position);
            }
            ImageStyle::Person => {
                // If `style` is set to `person`, apply an ellipse clipping to the image.
                // Assuming the image is square, it'll result in a circle.
                // But if the source image isn't square, it'll result in an ellipse, which is fine.
                let mut image_mut = image.borrow_mut();
                copy_image_within_ellipse(&source_image, &mut image_mut, final_position);
            }
        }

        Ok(())
    }
}

fn load_error_image() -> RgbaImage {
    const SIZE: u32 = 10;
    let mut image = RgbaImage::from_pixel(SIZE, SIZE, image::Rgba([255, 255, 255, 255]));
    for i in 0..SIZE {
        image.put_pixel(i, i, image::Rgba([0, 0, 0, 255]));
        image.put_pixel(i, SIZE - 1 - i, image::Rgba([0, 0, 0, 255]));
    }
    image
}

fn load_source_image(
    image: &adaptive_cards::Image<ElementLayoutData>,
    context: &LayoutContext<'_, '_, '_>,
    absolute_rect: Option<Rect>,
) -> RgbaImage {
    let source_image_size = SourceImageSize::from_image(image);

    let url = match source_image_size {
        SourceImageSize::Dynamic => {
            let Some(absolute_rect) = absolute_rect else {
                panic!("absolute_rect is required when a dynamically sized image is present");
            };

            let mut url = image.url.clone();
            url = url.replace("$width", &absolute_rect.width().to_string());
            url = url.replace("$height", &absolute_rect.height().to_string());
            url
        }
        SourceImageSize::Fixed => image.url.clone(),
    };

    let maybe_image = context
        .host_context
        .load_image_from_url(&url, image.body.as_ref());

    match maybe_image {
        Err(e) => {
            context
                .host_context
                .log_warn(&format!("Failed to load image from URL: {}", url));

            context.host_context.log_warn(&format!(" {}", e.message));

            for inner in e.inner {
                context.host_context.log_warn(&format!(" {}", inner));
            }

            load_error_image()
        }
        Ok(image) => image,
    }
}

fn copy_image(source: &RgbaImage, target: &mut MaskedImage, position: Point<i32>) {
    let x = position.x;
    let y = position.y;
    for sy in 0..source.height() {
        for sx in 0..source.width() {
            let px = source.get_pixel(sx, sy);
            let tx = x + sx.clamp_to_i32();
            let ty = y + sy.clamp_to_i32();
            let Ok(tx) = u32::try_from(tx) else {
                continue;
            };
            let Ok(ty) = u32::try_from(ty) else {
                continue;
            };
            target.draw_pixel(tx, ty, *px);
        }
    }
}

fn copy_image_within_ellipse(source: &RgbaImage, target: &mut MaskedImage, position: Point<i32>) {
    let x = position.x;
    let y = position.y;
    let sw = source.width() as f32;
    let sh = source.height() as f32;
    let cx = sw / 2.0;
    let cy = sh / 2.0;
    let rx = cx;
    let ry = cy;

    for sy in 0..source.height() {
        for sx in 0..source.width() {
            let dx = (sx as f32 - cx) / rx;
            let dy = (sy as f32 - cy) / ry;
            let dist = dx * dx + dy * dy;

            // Dist == 1.0 is on the ellipse boundary, < 1.0 is inside.
            if dist <= 1.0 {
                let px = source.get_pixel(sx, sy);
                let alpha = if dist > 0.9 {
                    // Blend near the edge for a smooth fade.
                    let blend_factor = 1.0 - ((dist - 0.9) / 0.1);
                    (px[3] as f32 * blend_factor) as u8
                } else {
                    px[3]
                };

                let tx = x + sx.clamp_to_i32();
                let ty = y + sy.clamp_to_i32();

                let Ok(tx) = u32::try_from(tx) else {
                    continue;
                };
                let Ok(ty) = u32::try_from(ty) else {
                    continue;
                };

                if tx < target.width() && ty < target.height() {
                    let mut tgt_px = target.get_pixel(tx, ty);
                    let inv_alpha = 255 - alpha;

                    // Simple alpha blending
                    for i in 0..3 {
                        let src_val = px[i] as u16;
                        let tgt_val = tgt_px[i] as u16;
                        let blended = (src_val * alpha as u16 + tgt_val * inv_alpha as u16) / 255;
                        tgt_px[i] = blended as u8;
                    }
                    tgt_px[3] = 255; // Fully opaque result
                    target.draw_pixel(tx, ty, tgt_px);
                }
            }
        }
    }
}

fn get_new_absolute_position(
    absolute_rect: Rect,
    source_image: &RgbaImage,
    horizontal_alignment: adaptive_cards::HorizontalAlignment,
) -> Point<i32> {
    match horizontal_alignment {
        adaptive_cards::HorizontalAlignment::Center => {
            let left = absolute_rect.left()
                + (absolute_rect.width().clamp_to_i32() - source_image.width().clamp_to_i32()) / 2;
            let top = absolute_rect.top()
                + (absolute_rect.height().clamp_to_i32() - source_image.height().clamp_to_i32())
                    / 2;
            Point::new(left, top)
        }
        adaptive_cards::HorizontalAlignment::Left => {
            Point::new(absolute_rect.left(), absolute_rect.top())
        }
        adaptive_cards::HorizontalAlignment::Right => {
            let left = absolute_rect.right() - source_image.width().clamp_to_i32();
            let top = absolute_rect.top();
            Point::new(left, top)
        }
    }
}

fn scale_to_fit_width(
    source_image_width: u32,
    target_width: u32,
    source_image_height: u32,
    source_image: &mut image::ImageBuffer<image::Rgba<u8>, Vec<u8>>,
) {
    if source_image_width != target_width {
        let target_height =
            (source_image_height as f32 * (target_width as f32 / source_image_width as f32)) as u32;
        *source_image = scale_to_fit_rect(&*source_image, target_width, target_height);
    }
}

fn scale_to_fit_rect(img: &RgbaImage, max_width: u32, max_height: u32) -> RgbaImage {
    let (width, height) = img.dimensions();
    let scale = (max_width as f32 / width as f32).min(max_height as f32 / height as f32);
    let new_width = (width as f32 * scale) as u32;
    let new_height = (height as f32 * scale) as u32;
    image::imageops::resize(img, new_width, new_height, FilterType::CatmullRom)
}
