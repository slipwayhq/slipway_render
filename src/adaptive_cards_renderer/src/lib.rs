// Allow unused code
#![allow(dead_code)]

mod debug_mode;
mod element_layout_data;
mod errors;
pub mod host_config_utils;
pub mod host_context;
mod layout_context;
mod layout_impl;
mod layout_scratch;
mod layoutable;
mod masked_image;
mod print_tree;
mod rect;
pub mod render;
mod utils;

pub use debug_mode::DebugMode;
pub use element_layout_data::ElementLayoutData;
pub use host_config_utils::default_host_config::default_host_config;
use image::Rgba;

// Bring in the `impl_as_trait!` macro.
extern crate adaptive_cards;

const TRANSPARENT: Rgba<u8> = Rgba::<u8>([0, 0, 0, 0]);

fn is_transparent(color: Rgba<u8>) -> bool {
    color.0[3] == 0
}
