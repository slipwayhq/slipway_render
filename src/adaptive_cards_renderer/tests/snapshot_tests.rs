use std::path::Path;

use adaptive_cards_host_config::HostConfig;
use adaptive_cards_renderer::{
    default_host_config,
    host_context::{HostContext, ResolvedFont},
    render::render_from_str,
    DebugMode,
};
use image::RgbaImage;

const CARD_EXTENSION: &str = ".card.json";

#[test]
fn snapshots() {
    // Get test name from SNAPSHOT_TEST_NAME environment variable.
    let search_path = std::env::var("SNAPSHOT_TEST_NAME")
        .map(|name| format!("snapshot_inputs/{}{}", name, CARD_EXTENSION))
        .unwrap_or_else(|_| format!("snapshot_inputs/*{}", CARD_EXTENSION));

    let debug_mode = match std::env::var("SNAPSHOT_DEBUG_MODE").as_deref() {
        Ok("full") => DebugMode::full(),
        Ok("outlines") => DebugMode::with_outlines(),
        Ok("transparent_masks") => DebugMode::with_transparent_masks(),
        _ => DebugMode::none(),
    };

    insta::with_settings!({}, {
        insta::glob!(&search_path, |path| {
            let path_string = path.to_string_lossy();
            let card_prefix = path_string.strip_suffix(CARD_EXTENSION).unwrap();

            let spec = load_spec_file_for(card_prefix);

            println!("Running snapshot test for {}", card_prefix);

            let json_data = std::fs::read_to_string(path).unwrap();
            let (image, card) = render_from_str(
                &json_data,
                &spec.host_config.unwrap_or_else(default_host_config),
                &MockHostContext {},
                spec.width,
                spec.height,
                debug_mode,
            )
            .unwrap();

            write_image_for(card_prefix, &image);

            insta::assert_json_snapshot!(card);
        });
    });
}

fn write_image_for(card_prefix: &str, image: &RgbaImage) {
    let image_path = format!("{}{}", card_prefix, ".png");
    image.save(image_path).unwrap();
}

fn load_spec_file_for(card_prefix: &str) -> SnapshotTestSpec {
    let spec_file_string = format!("{}{}", card_prefix, ".spec.json");
    let spec_file_path = Path::new(&spec_file_string);
    if spec_file_path.exists() {
        serde_json::from_str::<SnapshotTestSpec>(&std::fs::read_to_string(spec_file_path).unwrap())
            .unwrap()
    } else {
        SnapshotTestSpec {
            width: 1000,
            height: 800,
            host_config: None,
        }
    }
}

#[derive(Debug, serde::Deserialize)]
struct SnapshotTestSpec {
    width: u32,
    height: u32,
    host_config: Option<HostConfig>,
}

// For our tests we're always going to use Roboto, so the results are consistent across platforms.
const ROBOTO_TTF: &[u8] = include_bytes!("../../../fonts/Roboto.ttf");
const ROBOTO_MONO_TTF: &[u8] = include_bytes!("../../../fonts/RobotoMono.ttf");
const AIRPLANE_PNG: &[u8] =
    include_bytes!("../../../src/adaptive_cards_renderer/tests/assets/airplane.png");
struct MockHostContext {}
impl HostContext for MockHostContext {
    fn try_resolve_font(&self, family: &str) -> Option<ResolvedFont> {
        if family.to_lowercase().contains("mono") {
            Some(ResolvedFont {
                family: "Roboto Mono".to_string(),
                data: ROBOTO_MONO_TTF.to_vec(),
            })
        } else {
            Some(ResolvedFont {
                family: "Roboto".to_string(),
                data: ROBOTO_TTF.to_vec(),
            })
        }
    }

    fn run_callout(
        &self,
        _handle: &str,
        _input: &serde_json::Value,
    ) -> Result<RgbaImage, adaptive_cards_renderer::host_context::ComponentError> {
        unimplemented!()
    }

    fn load_image_from_url(
        &self,
        url: &str,
    ) -> Result<RgbaImage, adaptive_cards_renderer::host_context::ComponentError> {
        match url {
            "https://adaptivecards.io/content/airplane.png" => {
                Ok(image::load_from_memory(AIRPLANE_PNG).unwrap().to_rgba8())
            }
            _ => Err(adaptive_cards_renderer::host_context::ComponentError {
                message: format!("Image not found: {}", url),
            }),
        }
    }

    fn warn(&self, message: &str) {
        println!("Warning: {}", message);
    }
}
