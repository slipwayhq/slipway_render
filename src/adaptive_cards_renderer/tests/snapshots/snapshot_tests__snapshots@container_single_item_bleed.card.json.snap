---
source: adaptive_cards_renderer/tests/snapshot_tests.rs
expression: card
input_file: adaptive_cards_renderer/tests/snapshot_inputs/container_single_item_bleed.card.json
---
{
  "$schema": "../../../../adaptive_cards_data/schema/adaptive-card.schema.json",
  "body": [
    {
      "type": "Container",
      "bleed": true,
      "height": "stretch",
      "items": [
        {
          "type": "TextBlock",
          "text": "1",
          ".layout": {
            "rect": {
              "x": 17,
              "y": 17,
              "width": 116,
              "height": 10
            },
            "placement": "SoleVertical"
          }
        }
      ],
      "style": "accent",
      ".layout": {
        "rect": {
          "x": 0,
          "y": 0,
          "width": 150,
          "height": 150
        },
        "placement": "SoleVertical"
      }
    }
  ],
  "version": "1.5",
  "type": "AdaptiveCard",
  ".layout": {
    "rect": {
      "x": 0,
      "y": 0,
      "width": 150,
      "height": 150
    }
  }
}
